### PR TITLE
Prune old rows in `device_lists_changes_in_room` table.

### LINKED ISSUE #13043: Clear out `device_lists_changes_in_room` table periodically.
This issue has been migrated from [#13043](https://github.com/matrix-org/synapse/issues/13043).

---

We added the `device_lists_changes_in_room` table in #12321, we should clear it out periodically so that it doesn't grow indefinitely.

Note that we want to use this new table in e.g. `/sync` #12388, so we need to make sure we correctly handle that case.

### PR DESCRIPTION
**Pull request by erikjohnston**
Fixes #13043

The usages of the table mostly already correctly handled if we don't have old entries, as that was needed when we first added the table.

I arbitrarily set the prune time to 30 days. The only use for old entries is for sync streams that haven't synced since then, and we should very rarely see sync streams that haven't been used in 30 days.

Reviewable commit-by-commit.

### CODE REVIEW

File: synapse/storage/databases/main/devices.py [hunk 1]
```diff
@@ -2401,6 +2457,119 @@ async def set_device_change_last_converted_pos(
...
+        prune_before_ts = (
+            self.clock.time_msec() - PRUNE_DEVICE_LISTS_CHANGES_IN_ROOM_AGE.as_millis()
+        )
+
+        # Get stream ID corresponding to the prune_before_ts timestamp. We can
+        # delete all rows with a stream ID less than or equal to this, as they
+        # will be older than the cutoff.
+        #
+        # Some rows will have a NULL inserted_ts, but we can assume that the
+        # timestamp will montonically increase with stream ID, so we can safely
```
Comment (reivilibre) [inline synapse/storage/databases/main/devices.py]:
```suggestion
        # timestamp will monotonically increase with stream ID, so we can safely
```

File: synapse/storage/databases/main/devices.py [hunk 2]
```diff
@@ -1808,7 +1854,7 @@ def _get_all_device_list_changes_txn(
 
     async def get_device_list_changes_in_room(
         self, room_id: str, min_stream_id: int
-    ) -> Collection[tuple[str, str]]:
+    ) -> Collection[tuple[str, str]] | None:
```
Comment (reivilibre) [inline synapse/storage/databases/main/devices.py:1873]: Return docstring should be updated to say what this means

File: synapse/storage/databases/main/devices.py [hunk 3]
```diff
@@ -2401,6 +2457,119 @@ async def set_device_change_last_converted_pos(
             desc="set_device_change_last_converted_pos",
         )
 
+    @wrap_as_background_process("prune_device_lists_changes_in_room")
+    async def _prune_device_lists_changes_in_room(self) -> None:
+        """Delete old entries out of the `device_lists_changes_in_room`, so that
+        we don't
```
Comment (reivilibre) [inline synapse/storage/databases/main/devices.py]: so we don't.....? :)

File: synapse/storage/databases/main/devices.py [hunk 4]
```diff
@@ -2401,6 +2457,119 @@ async def set_device_change_last_converted_pos(
...
+        # We also delete in batches to avoid massive churn when initially
+        # clearing out all the old entries.
+        delete_sql = """
+            DELETE FROM device_lists_changes_in_room
+            WHERE stream_id IN (
+                SELECT stream_id FROM device_lists_changes_in_room
+                WHERE stream_id <= ?
+                ORDER BY stream_id ASC
+                LIMIT ?
+            )
```
Comment (reivilibre) [inline synapse/storage/databases/main/devices.py]: This will be inefficient for large deletions because it will necessitate keeping reconsidering dead tuples for deletion. Every subsequent deletion will take longer and longer, at least until vacuum kicks in.

Would be better if we added a `? < stream_id` bound on whatever we've already deleted before (I know it's a faff because you have to figure out what you deleted)

From memory a `WITH` expression might be the least painful way to get this.

File: synapse/storage/databases/main/devices.py [hunk 5]
```diff
@@ -2401,6 +2457,119 @@ async def set_device_change_last_converted_pos(
...
+            # Make sure to invalidate the cache of the minimum stream ID after
+            # deleting.
+            self._invalidate_cache_and_stream(
+                txn, self._get_min_device_lists_changes_in_room, keys=()
+            )
+
+            return num_deleted
+
+        num_rows_deleted = 0
+        while True:
```
Comment (reivilibre) [inline synapse/storage/databases/main/devices.py:2589]: Are we happy that there isn't much in the way of rate control on these deletions?

File: synapse/storage/databases/main/devices.py [hunk 6]
```diff
@@ -2459,6 +2628,14 @@ def __init__(
             columns=["room_id", "stream_id"],
         )
 
+        # Add indexes to speed up pruning of device_lists_changes_in_room
+        self.db_pool.updates.register_background_index_update(
+            BG_UPDATE_ADD_INSERTED_TS_INDEX,
+            index_name="device_lists_changes_in_room_inserted_ts_idx",
+            table="device_lists_changes_in_room",
+            columns=["inserted_ts"],
```
Comment (reivilibre) [inline synapse/storage/databases/main/devices.py:2679]: It seems it might be slightly preferable for this to be a partial index on `WHERE inserted_ts IS NOT NULL`.

File: synapse/storage/schema/main/delta/93/05_device_lists_room_timestamp.sql [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+-- License, or (at your option) any later version.
+--
+-- See the GNU Affero General Public License for more details:
+-- <https://www.gnu.org/licenses/agpl-3.0.html>.
+
+ALTER TABLE device_lists_changes_in_room ADD COLUMN inserted_ts BIGINT;
+
+-- Add a background update to add index
+INSERT INTO background_updates (ordering, update_name, progress_json) VALUES
+  (9005, 'device_lists_changes_in_room_inserted_ts_idx', '{}');
```
Comment (reivilibre) [inline synapse/storage/schema/main/delta/93/05_device_lists_room_timestamp.sql]:
```suggestion
  (9305, 'device_lists_changes_in_room_inserted_ts_idx', '{}');
```

File: synapse/storage/databases/main/devices.py [hunk 8]
```diff
@@ -2401,6 +2457,119 @@ async def set_device_change_last_converted_pos(
...
+
+        prune_before_ts = (
+            self.clock.time_msec() - PRUNE_DEVICE_LISTS_CHANGES_IN_ROOM_AGE.as_millis()
+        )
+
+        # Get stream ID corresponding to the prune_before_ts timestamp. We can
+        # delete all rows with a stream ID less than or equal to this, as they
+        # will be older than the cutoff.
+        #
+        # Some rows will have a NULL inserted_ts, but we can assume that the
```
Comment (reivilibre) [inline synapse/storage/databases/main/devices.py]: should briefly touch on the reason — that those rows existed before the column did

File: synapse/storage/databases/main/devices.py [hunk 9]
```diff
@@ -1143,6 +1160,35 @@ async def get_users_whose_devices_changed(
...
+    ) -> set[tuple[str, str]]:
+        """Get set of user/device ID tuple whose devices have changed since `from_key` that
+        are in the given list of user_ids.
+
+        Args:
+            from_key: The minimum device lists stream token to query device list changes for,
+                exclusive.
+            user_ids: If provided, only check if these users have changed their device lists.
+                Otherwise changes from all users are returned.
+            to_key: The maximum device lists stream token to query device list changes for,
```
Comment (reivilibre) [inline synapse/storage/databases/main/devices.py:1185]: should say what happens if None

**Review by reivilibre — CHANGES_REQUESTED** [review]
Structurally sane :)

File: synapse/storage/databases/main/devices.py (hunk 2 above)
Comment (erikjohnston) [reply synapse/storage/databases/main/devices.py:1873]: 9ee43bd61a

File: synapse/storage/databases/main/devices.py (hunk 3 above)
Comment (erikjohnston) [reply synapse/storage/databases/main/devices.py]: 017944472b

File: synapse/storage/databases/main/devices.py (hunk 5 above)
Comment (erikjohnston) [reply synapse/storage/databases/main/devices.py:2589]: de620fa

File: synapse/storage/databases/main/devices.py (hunk 4 above)
Comment (erikjohnston) [reply synapse/storage/databases/main/devices.py]: Mmmm, good point. Have added a limit that we keep track of when we're deleting lots and once 76e050b496

File: synapse/storage/databases/main/devices.py (hunk 6 above)
Comment (erikjohnston) [reply synapse/storage/databases/main/devices.py:2679]: Eh, we'll end up sooner rather than later having only non-null rows, so having the where clause just makes things a bit more complicated I think.

File: synapse/storage/databases/main/devices.py (hunk 8 above)
Comment (erikjohnston) [reply synapse/storage/databases/main/devices.py]: 484ca3a

File: synapse/storage/databases/main/devices.py (hunk 9 above)
Comment (erikjohnston) [reply synapse/storage/databases/main/devices.py:1185]: 484ca3a

File: synapse/storage/databases/main/devices.py (hunk 6 above)
Comment (reivilibre) [reply synapse/storage/databases/main/devices.py:2679]: The benefit would be that we don't get a lot of churn when first turning this on, since the index would start off empty.

Having a full index means we first have to build this index, filling it up with NULLs, before subsequently deleting all the rows it points to, leaving a lot of dead tuples in the index that have to be vacuumed eventually.

m.org currently has 5e9 rows; I think having a full index with just the nulls would give an index of 40-80 GB in size; which is not mega huge but it also won't get reclaimed except by a full vacuum afaiu.

The query already needs to have the WHERE clause so I don't really understand what you mean by more complicated; it looks like a 1-line change to me

File: synapse/storage/databases/main/devices.py [hunk 10]
```diff
@@ -2401,6 +2458,142 @@ async def set_device_change_last_converted_pos(
             desc="set_device_change_last_converted_pos",
         )
 
+    @wrap_as_background_process("prune_device_lists_changes_in_room")
+    async def _prune_device_lists_changes_in_room(self) -> None:
+        """Delete old entries out of the `device_lists_changes_in_room`, so that
+        we the table doesn't grow indefinitely.
```
Comment (reivilibre) [inline synapse/storage/databases/main/devices.py]:
```suggestion
        the table doesn't grow indefinitely.
```

3rd time lucky lol ? :D

File: synapse/storage/databases/main/devices.py [hunk 11]
```diff
@@ -1808,15 +1854,22 @@ def _get_all_device_list_changes_txn(
...
-            changed
+            changed, or None if the given stream ID is too old and so a complete
+            list cannot be calculated.
         """
 
+        lowest_known_stream_id = await self._get_min_device_lists_changes_in_room()
+
+        # Return early if there are no rows to process in device_lists_changes_in_room
+        if lowest_known_stream_id > min_stream_id:
+            return None
```
Comment (reivilibre) [inline synapse/storage/databases/main/devices.py]: There's a slim TOCTOU with this being non-transactional; this check should happen within the same transaction fetching the data, otherwise you risk someone getting an empty device list change set because the table got pruned at just the wrong moment.

File: synapse/storage/databases/main/devices.py [hunk 12]
```diff
@@ -2401,6 +2458,142 @@ async def set_device_change_last_converted_pos(
...
+        )
+
+        # Now delete all rows with stream_id less than the
+        # prune_before_stream_id.
+        #
+        # We also delete in batches to avoid massive churn when initially
+        # clearing out all the old entries.
+        #
+        # We set a minimum stream ID so that when we delete in batches the
+        # database doesn't have to scan through all the rows that were just
```
Comment (reivilibre) [inline synapse/storage/databases/main/devices.py]:
```suggestion
        # database doesn't have to scan through all the (dead) tuples that were just
```

minor, though Postgres distinguishes these and I think this terminology is the most likely to get you to the right part of the manual if you didn't understand this comment

File: synapse/handlers/device.py [hunk 13]
```diff
@@ -1193,7 +1194,16 @@ async def handle_room_un_partial_stated(self, room_id: str) -> None:
         changes = await self.store.get_device_list_changes_in_room(
             room_id, device_lists_stream_id
         )
-        local_changes = {(u, d) for u, d in changes if self.hs.is_mine_id(u)}
+        if changes is not None:
+            local_changes = {(u, d) for u, d in changes if self.hs.is_mine_id(u)}
+        else:
+            # The `device_lists_stream_id` is too old, so we need to fall back
```
Comment (reivilibre) [inline synapse/handlers/device.py:1200]: how much of a pain would it be to stand up a test against this case? I worry this is probably untested code. Being an edge case, it will therefore probably go unnoticed if it breaks until a very awkward moment. Would be nice to have *something*

File: synapse/storage/databases/main/devices.py [hunk 14]
```diff
@@ -2401,6 +2458,142 @@ async def set_device_change_last_converted_pos(
...
+            # We can't use rowcount as that is incorrect on SQLite when using
+            # RETURNING.
+            num_deleted = 0
+            for row in txn:
+                num_deleted += 1
+                min_stream_id = max(min_stream_id, row[0])
+
+            # Make sure to invalidate the cache of the minimum stream ID after
+            # deleting.
+            self._invalidate_cache_and_stream(
```
Comment (reivilibre) [inline synapse/storage/databases/main/devices.py]: Mrm, on that TOCTOU again, this being an asynchronous invalidation (because of workers), this is another hazard for giving the wrong answers.

I think after all we need to double-check in the transaction that reads the table.

Then again, is it worth having this cache if that's the case? Since it'd only be fast-exiting on the cold path, which seems a bit pointless.

**Review by reivilibre — CHANGES_REQUESTED** [review]
Seems like it's basically there, save for a bit of potential raciness and maybe a test?

File: synapse/storage/databases/main/devices.py (hunk 11 above)
Comment (erikjohnston) [reply synapse/storage/databases/main/devices.py]: Mmm, very good point. I had just copied existing code, but that doesn't actually work. I've updated it so that we run the check in the transaction and removed the cache 4712139ef9bbab843300a1116162e664395da699

File: synapse/storage/databases/main/devices.py (hunk 14 above)
Comment (erikjohnston) [reply synapse/storage/databases/main/devices.py]: 4712139ef9bbab843300a1116162e664395da699

File: synapse/storage/databases/main/devices.py (hunk 6 above)
Comment (erikjohnston) [reply synapse/storage/databases/main/devices.py:2679]: Mmm, fair on the churn point. 846491fd19

> The query already needs to have the WHERE clause so I don't really understand what you mean by more complicated; it looks like a 1-line change to me

It's not so much the mechanics but the conceptual overhead of having to think about whether the index can be used or not.

File: synapse/handlers/device.py (hunk 13 above)
Comment (erikjohnston) [reply synapse/handlers/device.py:1200]: Right, managed to wrangle a test bb6c559404

File: synapse/storage/databases/main/devices.py [hunk 15]
```diff
@@ -1748,35 +1788,39 @@ async def get_device_list_changes_in_rooms(
...
-        for chunk in batch_iter(changed_room_ids, 1000):
-            changes |= await self.db_pool.runInteraction(
-                "get_device_list_changes_in_rooms",
-                _get_device_list_changes_in_rooms_txn,
-                chunk,
-            )
+                args.append(from_token.stream)
+                args.append(to_token.get_max_stream_pos())
+
+                txn.execute(sql.format(clause=clause), args)
```
Comment (reivilibre) [inline synapse/storage/databases/main/devices.py]: what would you think about inlining the sql string here as an f-string and using `(*args, from_token.stream, to_token.get_max_stream_pos())` as the args instead of appending

I find it a little awkward to have to navigate up the diff to see the query (and then to piece the list together mentally); but maybe I bought into the 'locality of behaviour' cult too much.

File: tests/handlers/test_device.py [hunk 16]
```diff
@@ -581,3 +603,334 @@ def test_dehydrate_v2_and_fetch_events(self) -> None:
...
+            signing_key,
+        )
+        return FrozenEventV3(join_event_dict, room_version, {}, None)
+
+    @parameterized.expand([("not_pruned", False), ("pruned", True)])
+    @patch(
+        "synapse.storage.databases.main.devices.PRUNE_DEVICE_LISTS_CHANGES_IN_ROOM_AGE",
+        Duration(minutes=1),
+    )
+    def test_local_device_changes_sent_to_new_servers_on_un_partial_state(
```
Comment (reivilibre) [inline tests/handlers/test_device.py:769]: I imagine it was reasonably fiddly to get this test all pieced together; thanks for writing it!

**Review by reivilibre — APPROVED** [review]
Thanks!

**Comment by erikjohnston** [comment]
@reivilibre sorry, the worker integration tests raised an issue around the fact that taking the minimum stream ID of `device_lists_changes_in_room` is not a reliable way of determining the minimum safe stream ID. fee4c3f changes things so that we explicitly track in a dedicated table the maximum stream ID that we have pruned

File: synapse/storage/databases/main/devices.py [hunk 17]
```diff
@@ -2401,6 +2475,144 @@ async def set_device_change_last_converted_pos(
...
+
+        # The minimum stream ID to delete in the next batch, c.f. comment above.
+        # We default to 0 here as that is less than all possible stream IDs.
+        min_stream_id = 0
+
+        def prune_device_lists_changes_in_room_txn(txn: LoggingTransaction) -> int:
+            nonlocal min_stream_id
+
+            txn.execute(
+                delete_sql,
```
Comment (reivilibre) [inline synapse/storage/databases/main/devices.py:2575]: trying to read the last changed commit, I am realising this SQL would also be easier to read inlined (mostly to see the bound on `prune_before_stream_id`, i.e. keeping the params close the SQL is useful for legibility)

File: synapse/storage/schema/main/delta/94/04_device_lists_changes_max_pruned.sql [hunk 18]
```diff
@@ -0,0 +1,31 @@
...
+
+
+-- Tracks the maximum stream_id that has been deleted (pruned) from the
+-- device_lists_changes_in_room table. This is used to determine whether
+-- it's safe to read from that table for a given stream_id — if the
+-- requested stream_id is <= the value here, the data has been pruned and
+-- the table cannot provide a complete answer.
+--
+-- This replaces the previous approach of using MIN(stream_id) on the
+-- device_lists_changes_in_room table, which incorrectly returned 0 when
```
Comment (reivilibre) [inline synapse/storage/schema/main/delta/94/04_device_lists_changes_max_pruned.sql]: Ironically this new table is more confusing to me, as in I don't see why this is preventing anything.

By my reading, `COALESCE(MIN(stream_id), 0)` is the reason the previous approach returned 0 when empty. (So I think the comment here is maybe a bit misleading, or doesn't quite reveal the reason we need this?)

Trying to understand the difference, here is a (step by step) comparison table as I see it:

|                                                                 | Old `_get_min_device_lists_changes_in_room`                            | New `_get_max_pruned_device_lists_changes_in_room_txn`                               |
|-----------------------------------------------------------------|---------------------------------|------------------------------------|
| Seed (empty table)                                        | returns 0                       | returns 0                          |
| New rows added with stream_ids 2..=15 (empty table)             | returns 2                       | still returns 0                    |
| Prune with prune_before_stream_id 10                            | now returns MIN(stream_id) = 10 | now returns 10 (explicitly stored) |


|                                                                 | Old `_get_min_device_lists_changes_in_room`                            | New `_get_max_pruned_device_lists_changes_in_room_txn`                               |
|-----------------------------------------------------------------|---------------------------------|------------------------------------|
| Seed (populated table with stream_ids 2..=9)             | returns MIN(stream_id) = 2      | returns MIN(stream_id) - 1 = 1     |
| New rows added with stream_ids 10..=15 (populated table) | still returns 2                 | still returns 1                    |
| Prune with prune_before_stream_id 10                            | now returns MIN(stream_id) = 10 | now returns 10 (explicitly stored) |


Apart from the initial state before the first prune, it looks like both approaches behave the same between prunes?
What am I missing? Maybe the `device_lists_changes_in_room` rows get deleted as rooms get purged?

**Review by reivilibre — CHANGES_REQUESTED** [review]
Probably sane but would be good to have a really good idea of what the new table is for; it does feel like the kind of thing someone will dig up years down the line and have to scratch their head over it.

It doesn't sound insane so probably just a matter of getting the illustration right, e.g. with a step by step sequence to show what goes wrong with the 'simple' `MIN(stream_id)`

File: synapse/storage/schema/main/delta/94/04_device_lists_changes_max_pruned.sql (hunk 18 above)
Comment (erikjohnston) [reply synapse/storage/schema/main/delta/94/04_device_lists_changes_max_pruned.sql]: > By my reading, COALESCE(MIN(stream_id), 0) is the reason the previous approach returned 0 when empty. (So I think the comment here is maybe a bit misleading, or doesn't quite reveal the reason we need this?)

Argh, sorry that comment should have been deleted. I was messing around with testing getting LLM to generate the patch and it misunderstood the rationale (but got the change correct). I changed it locally but it got swallowed.

The actual problem (as per commit comment) comes from when we generate stream positions that *don't* have associated data in the `device_lists_changes_in_rooms` table, e.g. because the user isn't in any rooms. In that case if we insert a row into `device_lists_stream` table but *not* in  `device_lists_changes_in_rooms`, and then later one that inserts rows into *both*, then `MIN(stream_id)` will return a stream ID greater than the first row even though its data hasn't been pruned (there was just no associated data to fetch).

File: synapse/storage/databases/main/devices.py (hunk 17 above)
Comment (reivilibre) [reply synapse/storage/databases/main/devices.py:2575]: bump

File: synapse/storage/schema/main/delta/94/04_device_lists_changes_max_pruned.sql [hunk 19]
```diff
@@ -0,0 +1,27 @@
...
+-- See the GNU Affero General Public License for more details:
+-- <https://www.gnu.org/licenses/agpl-3.0.html>.
+
+
+-- Tracks the maximum stream_id that has been deleted (pruned) from the
+-- device_lists_changes_in_room table. This is used to determine whether
+-- it's safe to read from that table for a given stream_id — if the
+-- requested stream_id is < the value here, the data has been pruned and
+-- the table cannot provide a complete answer.
+CREATE TABLE IF NOT EXISTS device_lists_changes_in_room_max_pruned_stream_id (
```
Comment (reivilibre) [inline synapse/storage/schema/main/delta/94/04_device_lists_changes_max_pruned.sql:27]: would be good to still have a little distilled comment about why we need this table, which is that `device_lists_stream` somehow ties into this (I don't have a great picture of what this is otherwise I'd suggest some wording)

**Review by reivilibre — APPROVED** [review]
LGTM otherwise, thanks!

File: synapse/storage/schema/main/delta/94/04_device_lists_changes_max_pruned.sql (hunk 19 above)
Comment (erikjohnston) [reply synapse/storage/schema/main/delta/94/04_device_lists_changes_max_pruned.sql:27]: Does e49192d05a help?
